thoth/python/aiosource.py

```diff
@@ -0,0 +1,269 @@
+        url = self.url + "/" + package_name
+        links = []
+
+        _LOGGER.debug(f"Discovering package %r artifacts from %r", package_name, url)
+
+        async with aiohttp.ClientSession(raise_for_status=True) as session:
+            async with session.get(url) as response:
+                if response.status == 404:
+                    raise NotFound(f"Package {package_name} is not present on index {self.url} (index {self.name})")
+
+                text = await response.text()
+                soup = BeautifulSoup(text, "lxml")
+
+                links = soup.find_all("a")
+
+        artifacts = []
+        for link in links:
+            artifact_name = str(link["href"]).rsplit("/", maxsplit=1)
+            if len(artifact_name) == 2:
+                # If full URL provided by index.
+                artifact_name = artifact_name[1]
+            else:
+                artifact_name = artifact_name[0]
+
+            artifact_parts = artifact_name.rsplit("#", maxsplit=1)
+            if len(artifact_parts) == 2:
+                artifact_name = artifact_parts[0]
+
+            if not artifact_name.endswith((".tar.gz", ".whl")):
+                _LOGGER.debug("Link does not look like a package for %r: %r", package_name, link["href"])
+                continue
+
+            artifact_url = link["href"]
+            if not artifact_url.startswith(("http://", "https://")):
+                artifact_url = url + f"/{artifact_name}"
+
+            artifacts.append((artifact_name, artifact_url))
+
+        return AsyncIterableArtifacts(artifacts)
+
+    async def get_packages(self) -> AsyncIterablePackages:
+        """List packages available on the source package index."""
+        _LOGGER.debug(f"Discovering packages available on {self.url} (simple index name: {self.name})")
+
+        links = []
+
+        async with aiohttp.ClientSession(raise_for_status=True) as session:
+            async with session.get(self.url) as resp:
+                text = await resp.text()
+
+                soup = BeautifulSoup(text, "lxml")
+                links = soup.find_all("a")
+
+        if len(links) == 0:
+            return None
+
+        packages = set()
+        for link in links:
+            package_parts = link["href"].rsplit("/", maxsplit=2)
+            # According to PEP-503, package names must have trailing '/', but check this explicitly
+            if not package_parts[-1]:
+                package_parts = package_parts[:-1]
+            package_name = package_parts[-1]
+
+            # Discard links to parent dirs (package name of URL does not match the text.
+            link_text = link.text
+            if link_text.endswith("/"):
+                # Link names should end with trailing / according to PEEP:
+                #   https://www.python.org/dev/peps/pep-0503/
+                link_text = link_text[:-1]
+
+            if self.is_normalized_python_package_name(package_name) and package_name == link_text:
```

Comment:
`link_text` should be normalized, see https://github.com/thoth-station/python/pull/161

Reply:
syn?

Reply:
Any reason why not always operate on normalized package names?